crates/extensions/tedge-p11-server/src/server.rs

```diff
@@ -80,8 +84,115 @@ impl TedgeP11Server {
             }
         };
 
-        connection.write_frame(&response)?;
+        connection.write_frame(&response).context("write")?;
 
         Ok(())
     }
 }
+
+#[cfg(test)]
+mod tests {
+    use crate::client::TedgeP11Client;
+    use crate::service::*;
+    use std::io::Read;
+    use std::os::unix::net::UnixStream;
+    use std::time::Duration;
+
+    use super::*;
+
+    struct TestSigningService;
+
+    impl SigningService for TestSigningService {
+        fn choose_scheme(
+            &self,
+            _request: ChooseSchemeRequest,
+        ) -> anyhow::Result<ChooseSchemeResponse> {
+            Ok(ChooseSchemeResponse {
+                scheme: None,
+                algorithm: SignatureAlgorithm(rustls::SignatureAlgorithm::ECDSA),
+            })
+        }
+
+        fn sign(&self, _request: SignRequest) -> anyhow::Result<SignResponse> {
+            Ok(SignResponse(vec![]))
+        }
+    }
+
+    /// Check that client successfully receives responses from the server about the requests. Tests the
+    /// connection, framing, serialization, but not PKCS#11 layer itself.
+    #[tokio::test]
+    async fn server_works_with_client() {
+        let service = TestSigningService;
+        let server = TedgeP11Server::new(service).unwrap();
+        let tmpdir = tempfile::tempdir().unwrap();
+        let socket_path = tmpdir.path().join("test_socket.sock");
+        let listener = UnixListener::bind(&socket_path).unwrap();
+
+        tokio::spawn(async move { server.serve(listener).await });
+        // wait until the server calls accept()
+        tokio::time::sleep(Duration::from_millis(2)).await;
+
+        let client = TedgeP11Client {
+            socket_path: socket_path.into(),
+        };
+
+        tokio::task::spawn_blocking(move || {
+            client.choose_scheme(&[]).unwrap();
+            client.sign(&[]).unwrap();
```

Comment:
I would checks that some meaningful response is returned

Reply:
fixed in 970733c